### ISSUE TITLE
[Mailer] Delivery randomly fails with Google SMTP Relay

### ISSUE THREAD

**Issue by bpolaszek** [symfony/symfony#32148]
Hello there,

## TL;DR:
Using _Google SMTP Relay_ with the new _Mailer Component_ works in a random manner.

-----

### Background
I'm working on a new project for which I chose the new _Mailer component_ over _Swift Mailer_ with the same configuration of what I'm used to (i.e. using [Google SMTP Relay](https://support.google.com/a/answer/2956491)).

This service accessible from Google Apps allows using Google's SMTP servers **without authentication**, on an IP whitelist basis. 

The Google SMTP Relay DSN is the following one, and doesn't require any credentials:

```bash
smtp://smtp-relay.gmail.com:587?encryption=&auth_mode=
```

### Issue
I've been using it with SwiftMailer for years without any issue, but when using _Symfony Mailer_, it is _likely_ to result in the following exception:

> {"exception":"[object] (Symfony\Component\Mailer\Exception\TransportException(code: 0): Expected response code "250" but got an empty response. at ~/builds/prod/releases/4/vendor/symfony/mailer/Transport/Smtp/SmtpTransport.php:236)","command":"app:test","message":"Expected response code "250" but got an empty response."}

The _likely_ is the problem, because that exception doesn't occur every time.

When adding a `dump($response);` after this line: 
https://github.com/symfony/symfony/blob/aa4385dc125213342c8292c9060693fa51cbad4d/src/Symfony/Component/Mailer/Transport/Smtp/SmtpTransport.php#L142 

There's a 3/5 chance the dumped response will be 

```bash
"421 4.7.0 Try again later, closing connection. (EHLO) x12sm193387wrq.21 - gsmtp\r\n"
""
````

over

```bash
"""
250-smtp-relay.gmail.com at your service, [92.154.***.***]\r\n
250-SIZE 157286400\r\n
250-8BITMIME\r\n
250-STARTTLS\r\n
250-ENHANCEDSTATUSCODES\r\n
250-PIPELINING\r\n
250-CHUNKING\r\n
250 SMTPUTF8\r\n
"""
"250 2.1.0 OK f50sm190172eda.33 - gsmtp\r\n"
"250 2.1.5 OK f50sm190172eda.33 - gsmtp\r\n"
"354  Go ahead f50sm190172eda.33 - gsmtp\r\n"
"250 2.0.0 OK  1561367145 f50sm190172eda.33 - gsmtp\r\n"
```

You'll probably tell me the problem comes from Google, since it's a `421` response code from their servers. The problem is that the issue just _never_ occurs with  _SwiftMailer_. Does the new component does some more things (like some sort of handshaking or whatever - I have no knowledge of how this works) Google SMTP servers can't interpret properly?

### How to reproduce
When I run the following command several times:

```bash
# .env
###> symfony/mailer ###
MAILER_DSN=smtp://smtp-relay.gmail.com:587?encryption=&auth_mode=
###< symfony/mailer ###
###> symfony/swiftmailer-bundle ###
MAILER_URL=smtp://smtp-relay.gmail.com:587?encryption=&auth_mode=
###< symfony/swiftmailer-bundle ###
```

```php
namespace App\Command;

use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Console\Style\SymfonyStyle;
use Symfony\Component\Mailer\Exception\TransportExceptionInterface;
use Symfony\Component\Mailer\MailerInterface as SymfonyMailer;
use Symfony\Component\Mime\Email as SymfonyMailerEmail;
use Symfony\Component\Mime\Part\TextPart;

class TestCommand extends Command
{
    protected static $defaultName = 'app:test';

    /**
     * @var SymfonyMailer
     */
    private $symfonyMailer;
    
    /**
     * @var \Swift_Mailer
     */
    private $swiftMailer;

    public function __construct(SymfonyMailer $symfonyMailer, \Swift_Mailer $swiftMailer)
    {
        parent::__construct();
        $this->symfonyMailer = $symfonyMailer;
        $this->swiftMailer = $swiftMailer;
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {

        $io = new SymfonyStyle($input, $output);

        try {
            $io->comment('Sending through Symfony Mailer...');
            $message = new SymfonyMailerEmail();
            $message->setBody(new TextPart(\date(('Y-m-d H:i:s'))));
            $message->subject('Symfony Mailer');
            $message->addFrom('helloworld@mycustomdomain.com');
            $message->addTo('me@mycustomdomain.com');
            $this->symfonyMailer->send($message);
            $io->success('Done.');
        } catch (TransportExceptionInterface $e) {
            $io->error($e->getMessage());
        }

        try {
            $io->comment('Sending throught Swift Mailer...');
            $message = new \Swift_Message('Swift Mailer', \date(('Y-m-d H:i:s')));
            $message->addFrom('helloworld@mycustomdomain.com');
            $message->addTo('me@mycustomdomain.com');
            $this->swiftMailer->send($message);
            $io->success('Done.');
        } catch (\Exception $e) {
            $io->error($e->getMessage());
        }
    }
}
```

SwiftMailer _always_ succeeds, while the Mailer component succeeds _sometimes_.

Thank you,
Ben

**Comment by tienvx** [comment]
I can confirm this bug. This happen to me all of the times, not randomly.

Also, there is a question on stackoverflow: https://stackoverflow.com/questions/56633307/why-is-the-symfony-mailer-component-returning-an-empty-response

**Comment by fabpot** [comment]
@tienvx It happens to you with Gmail as well? With the same setup (no credentials but a whitelisted IP)?

**Comment by fabpot** [comment]
I've just spotted a difference between Swiftmailer and Symfony Mailer. It should not make any difference, but we never know with random failures. @tienvx Can you try #32894? It won't help with the general issue as whitelisted IP does not use authenticators at all.

**Comment by tienvx** [comment]
Let me try to create a reproduce project.

Link to the project: TBD

**Comment by fabpot** [comment]
One thing I don't understand is the "empty response part" as @bpolaszek mentioned that they get `421 4.7.0 Try again later, closing connection. (EHLO) x12sm193387wrq.21 - gsmtp` which should return response 421 instead. Can someone paste me the full output of the logs when that happens?

**Comment by fabpot** [comment]
The issue on Stackoverflow is different and easy to fix: the problem is that the `@` in the username must be encoded (can someone answers the question there, I don't have an account on SO).

**Comment by fabpot** [comment]
@tienvx Ok, I found the reason why we get a generic "empty response" exception. #32895 fixes that and you will now have the real reason for your issue (same for the SO issue where a proper error message will be displayed now). Can you apply the patch to see if you can now see the real reason for your issue?

**Comment by tienvx** [comment]
Wow that's quick! let me try

**Comment by fabpot** [comment]
@tienvx 220 is not an error code. What do you mean?

**Comment by tienvx** [comment]
I mean
```
Expected response code "220" but got an empty response.
```
If you have time, please try my reproduction project. All of the times, you will get that error

**Comment by fabpot** [comment]
@tienvx Your DSN looks wrong to me. It looks like you are using a Swiftmailer DSN, which is not the same as a DSN for Symfony Mailer. I would recommend you to install the `symfony/google-mailer` bridge and then set your username/password as env vars.

**Comment by fabpot** [comment]
@tienvx use `MAILER_DSN=smtp://tien.xuan.vo%40gmail.com:thisisyourpassword@smtp.gmail.com:465?encryption=ssl` if you really want to use a DSN.

**Comment by fabpot** [comment]
@bpolaszek Sorry for the noise on this issue as @tienvx issue was not related.

I've just made some changes to the component to ease debugging. Could you try the same code with Symfony 4.4-dev? And wrap your code like this so that we can have the full debug logs?

```php
try {
    $sent = $transport->send($email);
} catch (TransportExceptionInterface $e) {
    print $e->getDebug()."\n";

    throw $e;
}
if ($sent) {
    print $sent->getDebug()."\n";
}
```

**Comment by tienvx** [comment]
My problem is because of the wrong dsn. Use your dsn and it works. Sorry guys! my bad

**Comment by bpolaszek** [comment]
@fabpot I'm currently away from work with no 4G and a 56k-like DSL connection. I won't be able to test this before a few weeks 😕
I'll keep you in touch. 

Thank you for working on this!

**Comment by xabbuh** [comment]
@bpolaszek Do you have any news for us? :)

**Comment by bpolaszek** [comment]
Hello everyone, I'm back from vacations :-)

Good news, I just created a sample project with `symfony/mailer:4.3.4` and `symfony/google-mailer:4.3.4` and ran my example command: the messages get delivered! :rocket: 

Just 1 small issue remaining: when trying to send a message from a non-whitelisted IP address, here's the Swiftmailer console output (caught exception and $io->error(), cf my sample code at the top of the issue):

```
[ERROR] Expected response code "250" but got code "550", with message "550-5.7.1 Invalid credentials for relay         
                                                                        
                               n your G Suite SMTP Relay service doesn't match domain of
                               this email is being sent from. If you are trying to relay
                                 main that isn't registered under your G Suite account
                                     ope-from, you must configure your mail server
                                 SMTP AUTH to identify the sending domain or to present
                                      names in the HELO or EHLO command. For more
                                                                           
         550 5.7.1  https://support.google.com/a/answer/6140680#invalidcred i1sm130156edf.43 - gsmtp". 
```

And here's the output for Symfony Mailer:
```
PHP Fatal error:  Uncaught ErrorException: Notice: fwrite(): send of 6 bytes failed with errno=32 Broken pipe in /home/ben/PhpstormProjects/test-mailer/vendor/symfony/mailer/Transport/Smtp/Stream/AbstractStream.php:38
Stack trace:
#0 /home/ben/PhpstormProjects/test-mailer/vendor/symfony/mailer/Transport/Smtp/SmtpTransport.php(145): Symfony\Component\Mailer\Transport\Smtp\Stream\AbstractStream->write('QUIT\r\n')
#1 /home/ben/PhpstormProjects/test-mailer/vendor/symfony/mailer/Transport/Smtp/SmtpTransport.php(208): Symfony\Component\Mailer\Transport\Smtp\SmtpTransport->executeCommand('QUIT\r\n', Array)
#2 /home/ben/PhpstormProjects/test-mailer/vendor/symfony/mailer/Transport/Smtp/SmtpTransport.php(288): Symfony\Component\Mailer\Transport\Smtp\SmtpTransport->stop()
#3 [internal function]: Symfony\Component\Mailer\Transport\Smtp\SmtpTransport->__destruct()
#4 {main}
  thrown in /home/ben/PhpstormProjects/test-mailer/vendor/symfony/mailer/Transport/Smtp/Stream/AbstractStream.php on line 38
```

**Comment by fabpot** [comment]
@bpolaszek Can you try to see how we can catch the problem and throw the proper exception?

**Comment by bpolaszek** [comment]
@fabpot I'm currently quitting my actual employer and training my successor until the end of the week, before joining the famous Tilleuls Coop next Monday, so I can't promise anything for the moment 😐
Anyway, since the problem occurs when you're not allowed to use the SMTP relay, almost anyone can test the DSN and reproduce the exception. Otherwise, I'll try to have a look at this within the few weeks. I'll keep you in touch.

**Comment by bpolaszek** [comment]
@fabpot just had a quick look at the actual code:

https://github.com/symfony/symfony/blob/545d38a037098f0f21b11b380912151488acfac7/src/Symfony/Component/Mailer/Transport/Smtp/Stream/AbstractStream.php#L44-L49

`fwrite` returns `false` and this should be properly be converted to a `TransportException`. The problem is that `fwrite` doesn't just return `false`, it also emits a notice, which is converted to an `ErrorException`, preventing the `TransportException` to be thrown.

Should we just replace by a simple an efficient `@fwrite()` to do the trick?

**Comment by stof** [comment]
@bpolaszek `@fwrite` looks good here, as we already handle the error through the return value.
But would that give the detailed error message reported by Swiftmailer for this case ? Or would it should `Unable to write bytes on the wire.` ?

**Comment by bpolaszek** [comment]
@stof following the example code, here's the output for Swiftmailer when not whitelisted on the relay:
![Capture d’écran 2019-08-28 à 12 25 07](https://user-images.githubusercontent.com/5569077/63847823-e4a3f480-c98e-11e9-942b-ebe2a1b0b407.png)

The error comes at the very end when the pool is flushed.

**Comment by stof** [comment]
OK, so the swiftmailer case does not seem to show the error message reported by you previously either. Where does this message come from ?

**Comment by bpolaszek** [comment]
https://github.com/symfony/swiftmailer-bundle/blob/cb125b3648f132fb8070b55393f20cb310907d3b/EventListener/EmailSenderListener.php#L60-L66

Actually it does match the error I got with symfony/mailer at first.

**Comment by piotrecky** [comment]
Is the problem fixed? I still have a problem:

Expected response code 250 but got an empty response

**Comment by LuKePicci** [comment]
I'm still suffering this (not randomly) on 5.2. The `dump($response)` gives me the same:
> `"421 4.7.0 Try again later, closing connection. (EHLO) n18sm285960wrm.12 - gsmtp"`

and php error:
> request.CRITICAL: Uncaught PHP Exception Symfony\Component\Mailer\Exception\TransportException: "Expected response code "250" but got an empty response." at /home/site/vendor/symfony/mailer/Transport/Smtp/SmtpTransport.php line 300 {"exception":"[object] (Symfony\\Component\\Mailer\\Exception\\TransportException(code: 0): Expected response code \"250\" but got an empty response. at /home/site/vendor/symfony/mailer/Transport/Smtp/SmtpTransport.php:300)"} []

This is the MAILER_DSN (I also tried auth_mode=null and encryption=tls, no difference):
> `smtp://smtp-relay.gmail.com:587/?auth_mode=&encryption=ssl`

This command works just fine from the same machine running symfony:
```bash
curl --connect-timeout 15 -v smtp://smtp-relay.gmail.com:587 --mail-from contact@mydomain.com --mail-rcpt recipient@mydomain.com --ssl --upload-file message.txt
```

Edit: I managed to debug this way:
```php
    /**
     * Runs a command against the stream, expecting the given response codes.
     *
     * @param int[] $codes
     *
     * @return string The server response
     *
     * @throws TransportException when an invalid response if received
     *
     * @internal
     */
    public function executeCommand(string $command, array $codes): string
    {
        $this->stream->write($command);
dump($command);
        $response = $this->getFullResponse();
dump($response);
        $this->assertResponseCode($response, $codes);

        return $response;
    }
```

I see symfony is using an `EHLO [127.0.0.1]` command to begin the transaction, which seems to be invalid. I solved this by adding `&local_domain=buggy.wtf` at the and of my DSN.

**Comment by Myddna** [comment]
This is happening on Symfony 4.4 too. 

Tried to set local_domain as the authorized domain and as the mail domain too. We always get "empty response". We can send the emails using CURL, so it does not seem to be a configuration issue on the smtp-relay end.

Should we create a new issue?

**Comment by LuKePicci** [comment]
I think this should be simply reopened as the correct SMTP message is not actually being reported
________________________________

**Comment by derrabus** [comment]
@LuKePicci Can you please submit your issue as a new ticket? Only few people monitor bug reports that have been closed 1.5 years ago.

**Comment by MjeTjeu** [comment]
Maybe this helps:

[Laravel-incorrectly-handling-ehlo-for-smtp](https://laracasts.com/index.php/index.php/discuss/channels/laravel/laravel-incorrectly-handling-ehlo-for-smtp-from-artisancli-issues-with-gmails-smtp-relay-resolved)

**Comment by MjeTjeu** [comment]
Maybe this helps:

[incorrectly-handling-ehlo-for-smtp](https://laracasts.com/discuss/channels/laravel/laravel-incorrectly-handling-ehlo-for-smtp-from-artisancli-issues-with-gmails-smtp-relay-resolved)
